drgn_tools/locking.py

```diff
@@ -85,3 +229,238 @@ def for_each_mutex_waiter(prog: Program, mutex: Object) -> Iterable[Object]:
         prog.type("struct mutex_waiter"), mutex.wait_list.address_of_(), "list"
     ):
         yield waiter.task
+
+
+######################################
+# rwsem
+######################################
+
+# Masks for rwsem.count
+_RWSEM_WRITER_LOCKED = 1 << 0
+_RWSEM_FLAG_WAITERS = 1 << 1
+_RWSEM_FLAG_HANDOFF = 1 << 2
+# Bits 8-62(i.e. 55 bits of counter indicate number of current readers that hold the lock)
+_RWSEM_READER_MASK = 0x7FFFFFFFFFFFFF00  # Bits 8-62 - 55-bit reader count
+_RWSEM_WRITER_MASK = 1 << 0
+_RWSEM_READER_SHIFT = 8
+
+# Masks for rwsem.owner
+_RWSEM_READER_OWNED = 1 << 0
+_RWSEM_ANONYMOUSLY_OWNED = 1 << 0  # For old kernels
+_RWSEM_RD_NONSPINNABLE = 1 << 1
+_RWSEM_WR_NONSPINNABLE = 1 << 2
+
+# Linux kernel commit 617f3ef95177840c77f59c2aec1029d27d5547d6 ("locking/rwsem:
+# Remove reader optimistic spinning") (in v5.11) removed optimistic spinning for
+# readers and hence left one bit to check for spinnable
+_RWSEM_NONSPINNABLE = 1 << 1
+
+
+def rwsem_has_spinner(rwsem: Object) -> bool:
+    """
+    Check if rwsem has optimistic spinners or not
+
+    :param rwsem: ``struct rw_semaphore *``
+    :returns: True if rwsem has optimistic spinners, False otherwise.
+    """
+    return osq_is_locked(rwsem.osq.address_of_())
+
+
+def for_each_rwsem_waiter_entity(rwsem: Object) -> Iterable[Object]:
+    """
+    Find rwsem_waiter(s) for given rwsem
+
+    :param rwsem: ``struct rw_semaphore *``
+    :returns: Iterator of ``struct rwsem_waiter``
+    """
+
+    for waiter in list_for_each_entry(
+        "struct rwsem_waiter", rwsem.wait_list.address_of_(), "list"
+    ):
+        yield waiter
+
+
+def get_rwsem_owner(rwsem: Object) -> Object:
+    """
+    Find owner of  given rwsem
+
+    :param rwsem: ``struct rw_semaphore *``
+    :returns: ``struct task_struct *`` if owner can be found, NULL otherwise
+    """
+    prog = rwsem.prog_
+    if not rwsem.count.counter.value_():
+        print("rwsem is free.")
+        return NULL(prog, "struct task_struct *")
+
+    if is_rwsem_writer_owned(rwsem):
+        if (rwsem.owner.type_.type_name() != "atomic_long_t") and (
+            rwsem.owner.value_() & _RWSEM_ANONYMOUSLY_OWNED
+        ):
+            print("rwsem is owned by anonymous writer")
+            return NULL(prog, "struct task_struct *")
+        else:
+            owner = cast("struct task_struct *", rwsem.owner.counter)
+            return owner
+
+
+def get_rwsem_waiter_type(rwsem_waiter: Object) -> str:
+    """
+    Find type of an rwsem waiter
+
+    :param rwsem_waiter: ``struct rwsem_waiter``
+    :returns: str indicating type of rwsem waiter
+    """
+
+    prog = rwsem_waiter.prog_
+    if rwsem_waiter.type.value_() == prog["RWSEM_WAITING_FOR_WRITE"].value_():
+        waiter_type = "down_write"
+    elif rwsem_waiter.type.value_() == prog["RWSEM_WAITING_FOR_READ"].value_():
+        waiter_type = "down_read"
```

Comment:
Tiny nitpick, these can be done with `prog.constant()`, which I slightly prefer now that I've done the CTF / kallsyms implementation, as that allows us to skip the symbol table lookup and go straight into the hash table lookup for CTF enum values.

It's probably a negligible actual runtime impact, of course.